cli/util/uploader.go

```diff
@@ -0,0 +1,434 @@
+	}
+	defer p.Close()
+	net, err := p.NetworkInfo(ctx.Context, client.PrmNetworkInfo{})
+	if err != nil {
+		return cli.Exit(fmt.Errorf("failed to get network info: %w", err), 1)
+	}
+	homomorphicHashingDisabled := net.HomomorphicHashingDisabled()
+	lastMissingBlockIndex, err := fetchLatestMissingBlockIndex(ctx.Context, p, containerID, acc.PrivateKey(), attr, int(currentBlockHeight))
+	if err != nil {
+		return cli.Exit(fmt.Errorf("failed to fetch max block index from container: %w", err), 1)
+	}
+
+	fmt.Fprintln(ctx.App.Writer, "First block of latest incomplete batch uploaded to NeoFS container:", lastMissingBlockIndex)
+
+	if lastMissingBlockIndex > int(currentBlockHeight) {
+		fmt.Fprintf(ctx.App.Writer, "No new blocks to upload. Need to upload starting from %d, current height %d\n", lastMissingBlockIndex, currentBlockHeight)
+		return nil
+	}
+
+	for batchStart := lastMissingBlockIndex; batchStart <= int(currentBlockHeight); batchStart += searchBatchSize {
+		batchEnd := min(batchStart+searchBatchSize, int(currentBlockHeight)+1)
+		fmt.Fprintf(ctx.App.Writer, "Processing batch from %d to %d\n", batchStart, batchEnd)
+
+		errorChan := make(chan error)
+		doneChan := make(chan struct{})
+		var wg sync.WaitGroup
+		wg.Add(numWorkers)
+
+		for i := range numWorkers {
+			go func(i int) {
+				defer wg.Done()
+				for blockIndex := batchStart + i; blockIndex < batchEnd; blockIndex += numWorkers {
+					var blk *block.Block
+					err = retry(func() error {
+						blk, err = rpc.GetBlockByIndex(uint32(blockIndex))
+						if err != nil {
+							return fmt.Errorf("failed to fetch block %d: %w", blockIndex, err)
+						}
+						return nil
+					})
+					if err != nil {
+						select {
+						case errorChan <- err:
+						default:
+						}
+						return
+					}
+
+					bw := io.NewBufBinWriter()
+					blk.EncodeBinary(bw.BinWriter)
+					if bw.Err != nil {
+						errorChan <- fmt.Errorf("failed to encode block %d: %w", blockIndex, bw.Err)
+						return
+					}
+					attrs := []object.Attribute{
+						*object.NewAttribute(attr, strconv.Itoa(int(blk.Index))),
+						*object.NewAttribute("primary", strconv.Itoa(int(blk.PrimaryIndex))),
+						*object.NewAttribute("hash", blk.Hash().StringLE()),
+						*object.NewAttribute("prevHash", blk.PrevHash.StringLE()),
+						*object.NewAttribute("timestamp", strconv.FormatUint(blk.Timestamp, 10)),
+					}
+
+					err = retry(func() error {
+						return uploadObj(ctx.Context, p, signer, acc.PrivateKey().GetScriptHash(), containerID, bw.Bytes(), attrs, homomorphicHashingDisabled)
+					})
+					if err != nil {
+						select {
+						case errorChan <- err:
+						default:
+						}
+						return
+					}
+				}
+			}(i)
+		}
+
+		go func() {
+			wg.Wait()
+			close(doneChan)
+		}()
+
+		select {
+		case err := <-errorChan:
+			return cli.Exit(fmt.Errorf("upload error: %w", err), 1)
+		case <-doneChan:
+		}
+
+		fmt.Fprintf(ctx.App.Writer, "Successfully uploaded batch of blocks: from %d to %d\n", batchStart, batchEnd)
```

Comment:
s/batchEnd/batchEnd-1

Because we don't upload `batchEnd`-th block in this batch.